# Data Science Fundamentals: The Pandas Library
## Box Office Data

Today, we will be exploring fundamental concepts of pandas data manipulation to prepare a data set for modeling. 

__Data set:__ Information from IMDB about movies

__Our Goal:__ Process and clean the data to prepare it for modeling to predict the gross profit of a movie.

## Importing the packages that we'll need

One of the things that makes Python **great** for data science is all of the different libraries that exist so we don't have to code them from scratch. 
- [Numpy](https://numpy.org/) for scientific and mathematical computing
- [Pandas](https://pandas.pydata.org/) for data wrangling and analysis

In [1]:
# data analysis packages
import numpy as np
import pandas as pd

# default settings for pandas
pd.set_option('mode.chained_assignment', None)
pd.set_option('display.float_format', '{:,.2f}'.format)

## Import the data
Pandas can work with information from all kinds of data sources. Below, we'll import the data we need from a GitHub URL and read it into a Pandas Dataframe using the Pandas [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function.

In [2]:
# import data from github
data = pd.read_csv("https://github.com/autumntoney/predict_the_box_office/raw/master/movie_metadata.csv")

In [3]:
data 

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                  723.00    178.00   
1     Color     Gore Verbinski                  302.00    169.00   
2     Color         Sam Mendes                  602.00    148.00   
3     Color  Christopher Nolan                  813.00    164.00   
4       NaN        Doug Walker                     nan       nan   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                    1.00     87.00   
5039  Color                NaN                   43.00     43.00   
5040  Color   Benjamin Roberds                   13.00     76.00   
5041  Color        Daniel Hsia                   14.00    100.00   
5042  Color           Jon Gunn                   43.00     90.00   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                        0.00                  855.00  Joel David Moore   
1                      563.00                1,000.00     Orlando Bloom   
2                        0.00                  161.00      Rory Kinnear   
3                   22,000.00               23,000.00    Christian Bale   
4                      131.00                     nan        Rob Walker   
...                       ...                     ...               ...   
5038                     2.00                  318.00     Daphne Zuniga   
5039                      nan                  319.00     Valorie Curry   
5040                     0.00                    0.00     Maxwell Moody   
5041                     0.00                  489.00     Daniel Henney   
5042                    16.00                   16.00  Brian Herzlinger   

      actor_1_facebook_likes          gross                           genres  \
0                   1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1                  40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2                  11,000.00 200,074,175.00        Action|Adventure|Thriller   
3                  27,000.00 448,130,642.00                  Action|Thriller   
4                     131.00            nan                      Documentary   
...                      ...            ...                              ...   
5038                  637.00            nan                     Comedy|Drama   
5039                  841.00            nan     Crime|Drama|Mystery|Thriller   
5040                    0.00            nan            Drama|Horror|Thriller   
5041                  946.00      10,443.00             Comedy|Drama|Romance   
5042                   86.00      85,222.00                      Documentary   

      ... num_user_for_reviews language  country  content_rating  \
0     ...             3,054.00  English      USA           PG-13   
1     ...             1,238.00  English      USA           PG-13   
2     ...               994.00  English       UK           PG-13   
3     ...             2,701.00  English      USA           PG-13   
4     ...                  nan      NaN      NaN             NaN   
...   ...                  ...      ...      ...             ...   
5038  ...                 6.00  English   Canada             NaN   
5039  ...               359.00  English      USA           TV-14   
5040  ...                 3.00  English      USA             NaN   
5041  ...                 9.00  English      USA           PG-13   
5042  ...                84.00  English      USA              PG   

             budget  title_year actor_2_facebook_likes imdb_score  \
0    237,000,000.00    2,009.00                 936.00       7.90   
1    300,000,000.00    2,007.00               5,000.00       7.10   
2    245,000,000.00    2,015.00                 393.00       6.80   
3    250,000,000.00    2,012.00              23,000.00       8.50   
4               nan         nan                  12.00       7.10   
...             ...         ...                    ...        ...   
5038            nan    2,01

## Understand the data

In [4]:
# Check out the first lines of the data set
data.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                  723.00    178.00   
1  Color     Gore Verbinski                  302.00    169.00   
2  Color         Sam Mendes                  602.00    148.00   
3  Color  Christopher Nolan                  813.00    164.00   
4    NaN        Doug Walker                     nan       nan   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                     0.00                  855.00  Joel David Moore   
1                   563.00                1,000.00     Orlando Bloom   
2                     0.00                  161.00      Rory Kinnear   
3                22,000.00               23,000.00    Christian Bale   
4                   131.00                     nan        Rob Walker   

   actor_1_facebook_likes          gross                           genres  \
0                1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1               40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2               11,000.00 200,074,175.00        Action|Adventure|Thriller   
3               27,000.00 448,130,642.00                  Action|Thriller   
4                  131.00            nan                      Documentary   

   ... num_user_for_reviews language  country  content_rating         budget  \
0  ...             3,054.00  English      USA           PG-13 237,000,000.00   
1  ...             1,238.00  English      USA           PG-13 300,000,000.00   
2  ...               994.00  English       UK           PG-13 245,000,000.00   
3  ...             2,701.00  English      USA           PG-13 250,000,000.00   
4  ...                  nan      NaN      NaN             NaN            nan   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0    2,009.00                 936.00       7.90          1.78   
1    2,007.00               5,000.00       7.10          2.35   
2    2,015.00                 393.00       6.80          2.35   
3    2,012.00              23,000.00       8.50          2.35   
4         nan                  12.00       7.10           nan   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [5]:
# Checking the size of our data (rows, columns)
data.shape

(5043, 28)

In [6]:
# Get a concise summary of the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

In [7]:
# Understand the basic statistical details of the data set
data.describe()

num_critic_for_reviews  duration  director_facebook_likes  \
count                4,993.00  5,028.00                 4,939.00   
mean                   140.19    107.20                   686.51   
std                    121.60     25.20                 2,813.33   
min                      1.00      7.00                     0.00   
25%                     50.00     93.00                     7.00   
50%                    110.00    103.00                    49.00   
75%                    195.00    118.00                   194.50   
max                    813.00    511.00                23,000.00   

       actor_3_facebook_likes  actor_1_facebook_likes          gross  \
count                5,020.00                5,036.00       4,159.00   
mean                   645.01                6,560.05  48,507,385.63   
std                  1,665.04               15,020.76  68,471,915.43   
min                      0.00                    0.00         162.00   
25%                    133.00                  614.00   5,351,178.00   
50%                    371.50                  988.00  25,528,495.00   
75%                    636.00               11,000.00  62,319,957.00   
max                 23,000.00              640,000.00 760,505,847.00   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count         5,043.00                   5,043.00              5,030.00   
mean         83,668.16                   9,699.06                  1.37   
std         138,485.26                  18,163.80                  2.01   
min               5.00                       0.00                  0.00   
25%           8,593.50                   1,411.00                  0.00   
50%          34,359.00                   3,090.00                  1.00   
75%          96,309.00                  13,756.50                  2.00   
max       1,689,764.00                 656,730.00                 43.00   

       num_user_for_reviews            budget  title_year  \
count              5,022.00          4,551.00    4,935.00   
mean                 272.77     39,752,620.44    2,002.47   
std                  377.98    206,114,898.45       12.47   
min                    1.00            218.00    1,916.00   
25%                   65.00      6,000,000.00    1,999.00   
50%                  156.00     20,000,000.00    2,005.00   
75%                  326.00     45,000,000.00    2,011.00   
max                5,060.00 12,215,500,000.00    2,016.00   

       actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  
count                5,030.00    5,043.00      4,714.00              5,043.00  
mean                 1,651.75        6.44          2.22              7,525.96  
std                  4,042.44        1.13          1.39             19,320.45  
min                      0.00        1.60          1.18                  0.00  
25%                    281.00        5.80          1.85                  0.00  
50%                    595.00        6.60          2.35                166.00  
75%                    918.00        7.20          2.35              3,000.00  
max                137,000.00        9.50         16.00            349,000.00

__Missing Values:__ You can identify columns with missing values by looking at the `non-null` items in the `.info()` call, and the difference in number of data points in the `count` row of the `.describe()` function. For example, the `num_critic_for_reviews` column has a lower `count` than `duration`, so has missing values.

__Outliers:__ You can also identify if there are outliers in the data set by looking at the `.describe()` function's spread of data in comparison to the `min` and `max` values.  For example, the `duration` column's `max` value is 511, which would correspond to an 8.5 hour film. You might want to go back and check that the values for that film are valid, and correct or drop the values that aren't.  

NB: We don't have time to investigate outliers in this workshop.

## Clean the Data

Now that we understand the basics of what's in the data, we now need to clean the data before it's ready for modeling.  Things we'll cover:
- Duplicate data
- Missing data
- Manipulating data

NB: Most models can only process numerical data, so we will focus our cleaning on those columns.

In [8]:
# Rename columns to include units
data.rename(columns={'duration': 'duration_mins',
                    'budget': 'budget_usd',
                    'gross': 'gross_usd'}, inplace=True)
data.head()

color      director_name  num_critic_for_reviews  duration_mins  \
0  Color      James Cameron                  723.00         178.00   
1  Color     Gore Verbinski                  302.00         169.00   
2  Color         Sam Mendes                  602.00         148.00   
3  Color  Christopher Nolan                  813.00         164.00   
4    NaN        Doug Walker                     nan            nan   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                     0.00                  855.00  Joel David Moore   
1                   563.00                1,000.00     Orlando Bloom   
2                     0.00                  161.00      Rory Kinnear   
3                22,000.00               23,000.00    Christian Bale   
4                   131.00                     nan        Rob Walker   

   actor_1_facebook_likes      gross_usd                           genres  \
0                1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1               40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2               11,000.00 200,074,175.00        Action|Adventure|Thriller   
3               27,000.00 448,130,642.00                  Action|Thriller   
4                  131.00            nan                      Documentary   

   ... num_user_for_reviews language  country  content_rating     budget_usd  \
0  ...             3,054.00  English      USA           PG-13 237,000,000.00   
1  ...             1,238.00  English      USA           PG-13 300,000,000.00   
2  ...               994.00  English       UK           PG-13 245,000,000.00   
3  ...             2,701.00  English      USA           PG-13 250,000,000.00   
4  ...                  nan      NaN      NaN             NaN            nan   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0    2,009.00                 936.00       7.90          1.78   
1    2,007.00               5,000.00       7.10          2.35   
2    2,015.00                 393.00       6.80          2.35   
3    2,012.00              23,000.00       8.50          2.35   
4         nan                  12.00       7.10           nan   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

### Removing duplicates
Since our data includes that of movies, we will want to check for duplicated `movie_title`, but since there could be movie remakes in here as well, we should also check the `title_year` using the Pandas [`duplicated`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function.

In [9]:
# Checking to see what movies are duplicates.  
# Sorting by movie title to see duplicates
data[data.duplicated(subset=['title_year','movie_title'], keep=False)].sort_values('movie_title').head()

color      director_name  num_critic_for_reviews  duration_mins  \
3711  Color  Richard Fleischer                   69.00         127.00   
4894  Color  Richard Fleischer                   69.00         127.00   
4950  Color      David Hewlett                    8.00          88.00   
4949  Color      David Hewlett                    8.00          88.00   
1420  Color         Wes Craven                  256.00         101.00   

      director_facebook_likes  actor_3_facebook_likes     actor_2_name  \
3711                   130.00                   51.00  Robert J. Wilke   
4894                   130.00                   51.00  Robert J. Wilke   
4950                   686.00                  405.00    David Hewlett   
4949                   686.00                  405.00    David Hewlett   
1420                     0.00                  574.00        Lin Shaye   

      actor_1_facebook_likes     gross_usd  \
3711                  617.00           nan   
4894                  618.00           nan   
4950                  847.00           nan   
4949                  847.00           nan   
1420               40,000.00 26,505,000.00   

                                     genres  ... num_user_for_reviews  \
3711  Adventure|Drama|Family|Fantasy|Sci-Fi  ...               108.00   
4894  Adventure|Drama|Family|Fantasy|Sci-Fi  ...               108.00   
4950                                 Comedy  ...                46.00   
4949                                 Comedy  ...                46.00   
1420                                 Horror  ...               668.00   

     language  country  content_rating   budget_usd  title_year  \
3711  English      USA        Approved 5,000,000.00    1,954.00   
4894  English      USA        Approved 5,000,000.00    1,954.00   
4950  English   Canada             NaN   120,000.00    2,007.00   
4949  English   Canada             NaN   120,000.00    2,007.00   
1420  English      USA               X 1,800,000.00    1,984.00   

     actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
3711                  53.00       7.20          1.37                    0  
4894                  53.00       7.20          1.37                    0  
4950                 686.00       7.00          1.78                  377  
4949                 686.00       7.00          1.78                  377  
1420                 852.00       7.50          1.85                10000  

[5 rows x 28 columns]

We do have duplicated movies in our data.  We need to drop those duplicates to remove them from our data set using the Pandas [`drop_duplicates`](https://pandas.pydata.org/pandas-docs/version/0.17/generated/pandas.DataFrame.drop_duplicates.html) function.

In [10]:
# Drop all duplicate movie titles that were released in the same year
data = data.drop_duplicates(subset=['movie_title', 'title_year'], keep='first').copy()

In [11]:
data

color      director_name  num_critic_for_reviews  duration_mins  \
0     Color      James Cameron                  723.00         178.00   
1     Color     Gore Verbinski                  302.00         169.00   
2     Color         Sam Mendes                  602.00         148.00   
3     Color  Christopher Nolan                  813.00         164.00   
4       NaN        Doug Walker                     nan            nan   
...     ...                ...                     ...            ...   
5038  Color        Scott Smith                    1.00          87.00   
5039  Color                NaN                   43.00          43.00   
5040  Color   Benjamin Roberds                   13.00          76.00   
5041  Color        Daniel Hsia                   14.00         100.00   
5042  Color           Jon Gunn                   43.00          90.00   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                        0.00                  855.00  Joel David Moore   
1                      563.00                1,000.00     Orlando Bloom   
2                        0.00                  161.00      Rory Kinnear   
3                   22,000.00               23,000.00    Christian Bale   
4                      131.00                     nan        Rob Walker   
...                       ...                     ...               ...   
5038                     2.00                  318.00     Daphne Zuniga   
5039                      nan                  319.00     Valorie Curry   
5040                     0.00                    0.00     Maxwell Moody   
5041                     0.00                  489.00     Daniel Henney   
5042                    16.00                   16.00  Brian Herzlinger   

      actor_1_facebook_likes      gross_usd                           genres  \
0                   1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1                  40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2                  11,000.00 200,074,175.00        Action|Adventure|Thriller   
3                  27,000.00 448,130,642.00                  Action|Thriller   
4                     131.00            nan                      Documentary   
...                      ...            ...                              ...   
5038                  637.00            nan                     Comedy|Drama   
5039                  841.00            nan     Crime|Drama|Mystery|Thriller   
5040                    0.00            nan            Drama|Horror|Thriller   
5041                  946.00      10,443.00             Comedy|Drama|Romance   
5042                   86.00      85,222.00                      Documentary   

      ... num_user_for_reviews language  country  content_rating  \
0     ...             3,054.00  English      USA           PG-13   
1     ...             1,238.00  English      USA           PG-13   
2     ...               994.00  English       UK           PG-13   
3     ...             2,701.00  English      USA           PG-13   
4     ...                  nan      NaN      NaN             NaN   
...   ...                  ...      ...      ...             ...   
5038  ...                 6.00  English   Canada             NaN   
5039  ...               359.00  English      USA           TV-14   
5040  ...                 3.00  English      USA             NaN   
5041  ...                 9.00  English      USA           PG-13   
5042  ...                84.00  English      USA              PG   

         budget_usd  title_year actor_2_facebook_likes imdb_score  \
0    237,000,000.00    2,009.00                 936.00       7.90   
1    300,000,000.00    2,007.00               5,000.00       7.10   
2    245,000,000.00    2,015.00                 393.00       6.80   
3    250,000,000.00    2,012.00              23,000.00       8.50   
4               nan         nan                  12.00       7.10   
...             ...         ...     

### Missing values
Now that we've cleaned out the duplicates, let's take a look at the missing values.

In [12]:
# Show how many values are missing from each column
data.isna().sum()

color                         19
director_name                102
num_critic_for_reviews        49
duration_mins                 15
director_facebook_likes      102
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross_usd                    863
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                152
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               301
budget_usd                   485
title_year                   106
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 326
movie_facebook_likes           0
dtype: int64

In [13]:
data.isna()

color  director_name  num_critic_for_reviews  duration_mins  \
0     False          False                   False          False   
1     False          False                   False          False   
2     False          False                   False          False   
3     False          False                   False          False   
4      True          False                    True           True   
...     ...            ...                     ...            ...   
5038  False          False                   False          False   
5039  False           True                   False          False   
5040  False          False                   False          False   
5041  False          False                   False          False   
5042  False          False                   False          False   

      director_facebook_likes  actor_3_facebook_likes  actor_2_name  \
0                       False                   False         False   
1                       False                   False         False   
2                       False                   False         False   
3                       False                   False         False   
4                       False                    True         False   
...                       ...                     ...           ...   
5038                    False                   False         False   
5039                     True                   False         False   
5040                    False                   False         False   
5041                    False                   False         False   
5042                    False                   False         False   

      actor_1_facebook_likes  gross_usd  genres  ...  num_user_for_reviews  \
0                      False      False   False  ...                 False   
1                      False      False   False  ...                 False   
2                      False      False   False  ...                 False   
3                      False      False   False  ...                 False   
4                      False       True   False  ...                  True   
...                      ...        ...     ...  ...                   ...   
5038                   False       True   False  ...                 False   
5039                   False       True   False  ...                 False   
5040                   False       True   False  ...                 False   
5041                   False      False   False  ...                 False   
5042                   False      False   False  ...                 False   

      language  country  content_rating  budget_usd  title_year  \
0        False    False           False       False       False   
1        False    False           False       False       False   
2        False    False           False       False       False   
3        False    False           False       False       False   
4         True     True            True        True        True   
...        ...      ...             ...         ...         ...   
5038     False    False            True        True       False   
5039     False    False           False        True        True   
5040     False    False            True       False       False   
5041     False    False           False        True       False   
5042     False    False           False       False       False   

      actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  
0                      False       False         False                 False  
1                      False       False         False                 False  
2                      False       False         False                 False  
3                      False       False         False                 False  
4                      False       False          True                 False  
...                      ...         ...           ...                   ...  
5038                   False 

There are a good number of films missing the `aspect_ratio`, and using our knowledge of movies, it is unlikely that the aspect ratio is important in predicting the gross revenue of the film.  Let's drop the whole column using the Pandas [`drop`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) function.

In [14]:
# Drop the aspect ratio column,  axis=1 means drop the column
data.drop('aspect_ratio', axis=1, inplace=True)

In [15]:
data

color      director_name  num_critic_for_reviews  duration_mins  \
0     Color      James Cameron                  723.00         178.00   
1     Color     Gore Verbinski                  302.00         169.00   
2     Color         Sam Mendes                  602.00         148.00   
3     Color  Christopher Nolan                  813.00         164.00   
4       NaN        Doug Walker                     nan            nan   
...     ...                ...                     ...            ...   
5038  Color        Scott Smith                    1.00          87.00   
5039  Color                NaN                   43.00          43.00   
5040  Color   Benjamin Roberds                   13.00          76.00   
5041  Color        Daniel Hsia                   14.00         100.00   
5042  Color           Jon Gunn                   43.00          90.00   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                        0.00                  855.00  Joel David Moore   
1                      563.00                1,000.00     Orlando Bloom   
2                        0.00                  161.00      Rory Kinnear   
3                   22,000.00               23,000.00    Christian Bale   
4                      131.00                     nan        Rob Walker   
...                       ...                     ...               ...   
5038                     2.00                  318.00     Daphne Zuniga   
5039                      nan                  319.00     Valorie Curry   
5040                     0.00                    0.00     Maxwell Moody   
5041                     0.00                  489.00     Daniel Henney   
5042                    16.00                   16.00  Brian Herzlinger   

      actor_1_facebook_likes      gross_usd                           genres  \
0                   1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1                  40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2                  11,000.00 200,074,175.00        Action|Adventure|Thriller   
3                  27,000.00 448,130,642.00                  Action|Thriller   
4                     131.00            nan                      Documentary   
...                      ...            ...                              ...   
5038                  637.00            nan                     Comedy|Drama   
5039                  841.00            nan     Crime|Drama|Mystery|Thriller   
5040                    0.00            nan            Drama|Horror|Thriller   
5041                  946.00      10,443.00             Comedy|Drama|Romance   
5042                   86.00      85,222.00                      Documentary   

      ...                                    movie_imdb_link  \
0     ...  http://www.imdb.com/title/tt0499549/?ref_=fn_t...   
1     ...  http://www.imdb.com/title/tt0449088/?ref_=fn_t...   
2     ...  http://www.imdb.com/title/tt2379713/?ref_=fn_t...   
3     ...  http://www.imdb.com/title/tt1345836/?ref_=fn_t...   
4     ...  http://www.imdb.com/title/tt5289954/?ref_=fn_t...   
...   ...                                                ...   
5038  ...  http://www.imdb.com/title/tt3000844/?ref_=fn_t...   
5039  ...  http://www.imdb.com/title/tt2071645/?ref_=fn_t...   
5040  ...  http://www.imdb.com/title/tt2107644/?ref_=fn_t...   
5041  ...  http://www.imdb.com/title/tt2070597/?ref_=fn_t...   
5042  ...  http://www.imdb.com/title/tt0378407/?ref_=fn_t...   

     num_user_for_reviews  language  country content_rating     budget_usd  \
0                3,054.00   English      USA          PG-13 237,000,000.00   
1                1,238.00   English      USA          PG-13 300,000,000.00   
2                  994.00   English       UK          PG-13 245,000,000.00   
3                2,701.00   English      USA          PG-13 250,000,000.00   
4                     nan       NaN      NaN            NaN            nan   
...                   ...     

Now, since we're wanting to eventually predict the `gross` revenue of the movie, let's take a look at the missing values of `gross`

In [16]:
# What percent of the gross values are missing?
print(len(data[data['gross_usd'].isna()])/len(data))
# Or, more simply
data['gross_usd'].isnull().mean()

0.17544216304126856


0.17544216304126856

In [17]:
# show what movies are missing the gross values
data[data['gross_usd'].isna()].head()

color director_name  num_critic_for_reviews  duration_mins  \
4      NaN   Doug Walker                     nan            nan   
84   Color  Roland Joffé                   10.00         109.00   
98   Color  Hideaki Anno                    1.00         120.00   
177  Color           NaN                   21.00          60.00   
199  Color    Matt Birch                    1.00            nan   

     director_facebook_likes  actor_3_facebook_likes           actor_2_name  \
4                     131.00                     nan             Rob Walker   
84                    596.00                  283.00          Alice Englert   
98                     28.00                   12.00      Shin'ya Tsukamoto   
177                      nan                  184.00  Philip Michael Thomas   
199                     0.00                  159.00            Dave Legeno   

     actor_1_facebook_likes  gross_usd                                genres  \
4                    131.00        nan                           Documentary   
84                   622.00        nan       Action|Adventure|Romance|Sci-Fi   
98                   544.00        nan  Action|Adventure|Drama|Horror|Sci-Fi   
177                  982.00        nan   Action|Crime|Drama|Mystery|Thriller   
199               10,000.00        nan                        Action|Fantasy   

     ...                                    movie_imdb_link  \
4    ...  http://www.imdb.com/title/tt5289954/?ref_=fn_t...   
84   ...  http://www.imdb.com/title/tt1321869/?ref_=fn_t...   
98   ...  http://www.imdb.com/title/tt4262980/?ref_=fn_t...   
177  ...  http://www.imdb.com/title/tt0086759/?ref_=fn_t...   
199  ...  http://www.imdb.com/title/tt1680310/?ref_=fn_t...   

    num_user_for_reviews  language  country content_rating   budget_usd  \
4                    nan       NaN      NaN            NaN          nan   
84                 15.00   English  Belgium              R          nan   
98                 13.00  Japanese    Japan            NaN          nan   
177                74.00   English      USA          TV-14 1,500,000.00   
199                 2.00   English       UK            NaN          nan   

    title_year actor_2_facebook_likes  imdb_score movie_facebook_likes  
4          nan                  12.00        7.10                    0  
84    2,015.00                 525.00        4.50                  677  
98    2,016.00                 106.00        8.20                    0  
177        nan                 321.00        7.50                    0  
199   2,011.00                 570.00        7.50                   40  

[5 rows x 27 columns]

Since we eventually want to predict the gross revenue of a movie, having a null value for gross revenue will not help us train our model, so we should drop them using the Pandas [`dropna`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) function.

In [18]:
# Drop all null values of gross
# You might be interested in how the movie performed
# so don't want to impute these values (skew the analysis)
data.dropna(subset=['gross_usd'], how='all', inplace=True)
# Check the size of the data after dropping null values
data.shape

(4056, 27)

Check out the highest-grossing movies by sorting the data.

In [20]:
# Sort all of the values by gross
data['gross_usd'].sort_values(ascending=False)

0      760,505,847.00
26     658,672,302.00
29     652,177,271.00
17     623,279,547.00
66     533,316,061.00
            ...      
3552           828.00
3840           728.00
4606           721.00
4607           703.00
3330           162.00
Name: gross_usd, Length: 4056, dtype: float64

Knowing a bit about movies, it makes sense that the budget of a film might be important to predicting how much money the movie makes, so let's check out the missing budget values.

In [ ]:
# Identifying the percent of budget values that are missing
***ADD CODE HERE***

Since we have between 5-10% of budget values missing, we can impute (or fill in) the missing values without creating too much bias in the data. 

Using our knowledge of films, I know that film budgets in recent years has increased, so just imputing the mean or median budget for all missing values would not make much sense.  To get around this, we can find the median gross revenue for each year using the Pandas [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) function, and then impute based on those values using the [`fillna`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) function.

In [21]:
# Drop movies where the year is missing
data.dropna(subset=['title_year'], how='all', inplace=True)
# Convert all years to integers
data['title_year'] = data['title_year'].astype(int)

In [22]:
data

color      director_name  num_critic_for_reviews  duration_mins  \
0     Color      James Cameron                  723.00         178.00   
1     Color     Gore Verbinski                  302.00         169.00   
2     Color         Sam Mendes                  602.00         148.00   
3     Color  Christopher Nolan                  813.00         164.00   
5     Color     Andrew Stanton                  462.00         132.00   
...     ...                ...                     ...            ...   
5034  Color   Neill Dela Llana                   35.00          80.00   
5035  Color   Robert Rodriguez                   56.00          81.00   
5037  Color       Edward Burns                   14.00          95.00   
5041  Color        Daniel Hsia                   14.00         100.00   
5042  Color           Jon Gunn                   43.00          90.00   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                        0.00                  855.00    Joel David Moore   
1                      563.00                1,000.00       Orlando Bloom   
2                        0.00                  161.00        Rory Kinnear   
3                   22,000.00               23,000.00      Christian Bale   
5                      475.00                  530.00     Samantha Morton   
...                       ...                     ...                 ...   
5034                     0.00                    0.00     Edgar Tancangco   
5035                     0.00                    6.00     Peter Marquardt   
5037                     0.00                  133.00  Caitlin FitzGerald   
5041                     0.00                  489.00       Daniel Henney   
5042                    16.00                   16.00    Brian Herzlinger   

      actor_1_facebook_likes      gross_usd  \
0                   1,000.00 760,505,847.00   
1                  40,000.00 309,404,152.00   
2                  11,000.00 200,074,175.00   
3                  27,000.00 448,130,642.00   
5                     640.00  73,058,679.00   
...                      ...            ...   
5034                    0.00      70,071.00   
5035                  121.00   2,040,920.00   
5037                  296.00       4,584.00   
5041                  946.00      10,443.00   
5042                   86.00      85,222.00   

                                   genres  ...  \
0         Action|Adventure|Fantasy|Sci-Fi  ...   
1                Action|Adventure|Fantasy  ...   
2               Action|Adventure|Thriller  ...   
3                         Action|Thriller  ...   
5                 Action|Adventure|Sci-Fi  ...   
...                                   ...  ...   
5034                             Thriller  ...   
5035  Action|Crime|Drama|Romance|Thriller  ...   
5037                         Comedy|Drama  ...   
5041                 Comedy|Drama|Romance  ...   
5042                          Documentary  ...   

                                        movie_imdb_link num_user_for_reviews  \
0     http://www.imdb.com/title/tt0499549/?ref_=fn_t...             3,054.00   
1     http://www.imdb.com/title/tt0449088/?ref_=fn_t...             1,238.00   
2     http://www.imdb.com/title/tt2379713/?ref_=fn_t...               994.00   
3     http://www.imdb.com/title/tt1345836/?ref_=fn_t...             2,701.00   
5     http://www.imdb.com/title/tt0401729/?ref_=fn_t...               738.00   
...                                                 ...                  ...   
5034  http://www.imdb.com/title/tt0428303/?ref_=fn_t...                35.00   
5035  http://www.imdb.com/title/tt0104815/?ref_=fn_t...               130.00   
5037  http://www.imdb.com/title/tt1880418/?ref_=fn_t...                14.00   
5041  http://www.imdb.com/title/tt2070597/?ref_=fn_t...                 9.00   
5042  http://www.imdb.com/title/tt0378407/?ref_=fn_t...                84.00   

      language      country content_rating     budget_usd title_year  \
0    

In [24]:
data['title_year']

0       2009
1       2007
2       2015
3       2012
5       2012
        ... 
5034    2005
5035    1992
5037    2011
5041    2012
5042    2004
Name: title_year, Length: 4054, dtype: int64

In [25]:
# Calculate median budgets per year
# Impute the median budgets per year for missing budget data
data['budget_usd'] = data['budget_usd'].fillna(data.groupby('title_year')['budget_usd'].transform('median'))

In [28]:
data

color      director_name  num_critic_for_reviews  duration_mins  \
0     Color      James Cameron                  723.00         178.00   
1     Color     Gore Verbinski                  302.00         169.00   
2     Color         Sam Mendes                  602.00         148.00   
3     Color  Christopher Nolan                  813.00         164.00   
5     Color     Andrew Stanton                  462.00         132.00   
...     ...                ...                     ...            ...   
5034  Color   Neill Dela Llana                   35.00          80.00   
5035  Color   Robert Rodriguez                   56.00          81.00   
5037  Color       Edward Burns                   14.00          95.00   
5041  Color        Daniel Hsia                   14.00         100.00   
5042  Color           Jon Gunn                   43.00          90.00   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                        0.00                  855.00    Joel David Moore   
1                      563.00                1,000.00       Orlando Bloom   
2                        0.00                  161.00        Rory Kinnear   
3                   22,000.00               23,000.00      Christian Bale   
5                      475.00                  530.00     Samantha Morton   
...                       ...                     ...                 ...   
5034                     0.00                    0.00     Edgar Tancangco   
5035                     0.00                    6.00     Peter Marquardt   
5037                     0.00                  133.00  Caitlin FitzGerald   
5041                     0.00                  489.00       Daniel Henney   
5042                    16.00                   16.00    Brian Herzlinger   

      actor_1_facebook_likes      gross_usd  \
0                   1,000.00 760,505,847.00   
1                  40,000.00 309,404,152.00   
2                  11,000.00 200,074,175.00   
3                  27,000.00 448,130,642.00   
5                     640.00  73,058,679.00   
...                      ...            ...   
5034                    0.00      70,071.00   
5035                  121.00   2,040,920.00   
5037                  296.00       4,584.00   
5041                  946.00      10,443.00   
5042                   86.00      85,222.00   

                                   genres  ...  \
0         Action|Adventure|Fantasy|Sci-Fi  ...   
1                Action|Adventure|Fantasy  ...   
2               Action|Adventure|Thriller  ...   
3                         Action|Thriller  ...   
5                 Action|Adventure|Sci-Fi  ...   
...                                   ...  ...   
5034                             Thriller  ...   
5035  Action|Crime|Drama|Romance|Thriller  ...   
5037                         Comedy|Drama  ...   
5041                 Comedy|Drama|Romance  ...   
5042                          Documentary  ...   

                                        movie_imdb_link num_user_for_reviews  \
0     http://www.imdb.com/title/tt0499549/?ref_=fn_t...             3,054.00   
1     http://www.imdb.com/title/tt0449088/?ref_=fn_t...             1,238.00   
2     http://www.imdb.com/title/tt2379713/?ref_=fn_t...               994.00   
3     http://www.imdb.com/title/tt1345836/?ref_=fn_t...             2,701.00   
5     http://www.imdb.com/title/tt0401729/?ref_=fn_t...               738.00   
...                                                 ...                  ...   
5034  http://www.imdb.com/title/tt0428303/?ref_=fn_t...                35.00   
5035  http://www.imdb.com/title/tt0104815/?ref_=fn_t...               130.00   
5037  http://www.imdb.com/title/tt1880418/?ref_=fn_t...                14.00   
5041  http://www.imdb.com/title/tt2070597/?ref_=fn_t...                 9.00   
5042  http://www.imdb.com/title/tt0378407/?ref_=fn_t...                84.00   

      language      country content_rating     budget_usd title_year  \
0    

In [29]:
# Re-check to see if there are still any missing budget values
data['budget_usd'].isna().sum()

1

In [30]:
# What is the movie with missing budget still?
data[data['budget_usd'].isna()].head()

color director_name  num_critic_for_reviews  duration_mins  \
4647  Color   James Algar                  116.00          70.00   

      director_facebook_likes  actor_3_facebook_likes    actor_2_name  \
4647                    11.00                    8.00  Donnie Dunagan   

      actor_1_facebook_likes      gross_usd                  genres  ...  \
4647                   16.00 102,797,150.00  Animation|Drama|Family  ...   

                                        movie_imdb_link num_user_for_reviews  \
4647  http://www.imdb.com/title/tt0034492/?ref_=fn_t...               136.00   

      language  country content_rating  budget_usd title_year  \
4647   English      USA       Approved         nan       1942   

     actor_2_facebook_likes  imdb_score movie_facebook_likes  
4647                  12.00        7.40                    0  

[1 rows x 27 columns]

There are no other movies from the year 1942, so we cannot impute based on year.  Other options:
- Impute based on the overall median budget
- Drop the film from the data
- Impute based on other data

Since it's only one film, and for the sake of time, we will drop the film from the data set. 

In [31]:
data.dropna(subset=['budget_usd'], axis=0, inplace=True)

Now, let's take a look at the different countries of origin in the data.  There are likely differences in the gross revenue of movies from different countries.

In [32]:
# Identify all of the unique countries
data['country'].unique()

array(['USA', 'UK', 'New Zealand', 'Canada', 'Australia', 'Germany',
       'China', 'New Line', 'France', 'Japan', 'Spain', 'Hong Kong',
       'Czech Republic', 'India', 'Peru', 'South Korea', 'Aruba',
       'Denmark', 'Mexico', 'Belgium', 'Ireland', 'South Africa', 'Italy',
       'Romania', 'Chile', 'Netherlands', 'Hungary', 'Russia', 'Greece',
       'Taiwan', 'Official site', 'Thailand', 'Iran', 'West Germany',
       'Georgia', 'Iceland', 'Brazil', 'Finland', 'Norway', 'Sweden',
       'Argentina', 'Colombia', 'Poland', 'Israel', 'Indonesia',
       'Afghanistan', 'Cameroon', 'Philippines'], dtype=object)

In [33]:
# Find how many movies are in each country
counts = data['country'].value_counts()
counts

USA               3153
UK                 352
France             120
Germany             85
Canada              72
Australia           40
Spain               26
India               24
Japan               17
China               17
Hong Kong           15
Mexico              14
Italy               11
South Korea         10
Ireland             10
New Zealand         10
Denmark              9
Brazil               5
Thailand             5
Israel               4
Russia               4
Iran                 4
Norway               4
Czech Republic       3
Netherlands          3
Argentina            3
Sweden               3
South Africa         3
Hungary              2
Romania              2
Iceland              2
Poland               2
Belgium              2
Greece               2
Taiwan               2
Colombia             1
New Line             1
Peru                 1
Cameroon             1
Georgia              1
Aruba                1
Chile                1
Afghanistan          1
Indonesia  

Since most movies are from the US, UK, and France let's just focus on movies from those three countries, so the other countries don't skew our predictions.

In [34]:
# Select just the countries with the 3 largest number of films
counts.nlargest(3)

USA       3153
UK         352
France     120
Name: country, dtype: int64

In [35]:
# Select just the country names of the three largest
counts.nlargest(3).index

Index(['USA', 'UK', 'France'], dtype='object')

In [36]:
# Select the data from only the top 3 countries
data = data[data['country'].isin(counts.nlargest(3).index)].copy()

In [37]:
data

color      director_name  num_critic_for_reviews  duration_mins  \
0     Color      James Cameron                  723.00         178.00   
1     Color     Gore Verbinski                  302.00         169.00   
2     Color         Sam Mendes                  602.00         148.00   
3     Color  Christopher Nolan                  813.00         164.00   
5     Color     Andrew Stanton                  462.00         132.00   
...     ...                ...                     ...            ...   
5033  Color      Shane Carruth                  143.00          77.00   
5035  Color   Robert Rodriguez                   56.00          81.00   
5037  Color       Edward Burns                   14.00          95.00   
5041  Color        Daniel Hsia                   14.00         100.00   
5042  Color           Jon Gunn                   43.00          90.00   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                        0.00                  855.00    Joel David Moore   
1                      563.00                1,000.00       Orlando Bloom   
2                        0.00                  161.00        Rory Kinnear   
3                   22,000.00               23,000.00      Christian Bale   
5                      475.00                  530.00     Samantha Morton   
...                       ...                     ...                 ...   
5033                   291.00                    8.00      David Sullivan   
5035                     0.00                    6.00     Peter Marquardt   
5037                     0.00                  133.00  Caitlin FitzGerald   
5041                     0.00                  489.00       Daniel Henney   
5042                    16.00                   16.00    Brian Herzlinger   

      actor_1_facebook_likes      gross_usd  \
0                   1,000.00 760,505,847.00   
1                  40,000.00 309,404,152.00   
2                  11,000.00 200,074,175.00   
3                  27,000.00 448,130,642.00   
5                     640.00  73,058,679.00   
...                      ...            ...   
5033                  291.00     424,760.00   
5035                  121.00   2,040,920.00   
5037                  296.00       4,584.00   
5041                  946.00      10,443.00   
5042                   86.00      85,222.00   

                                   genres  ...  \
0         Action|Adventure|Fantasy|Sci-Fi  ...   
1                Action|Adventure|Fantasy  ...   
2               Action|Adventure|Thriller  ...   
3                         Action|Thriller  ...   
5                 Action|Adventure|Sci-Fi  ...   
...                                   ...  ...   
5033                Drama|Sci-Fi|Thriller  ...   
5035  Action|Crime|Drama|Romance|Thriller  ...   
5037                         Comedy|Drama  ...   
5041                 Comedy|Drama|Romance  ...   
5042                          Documentary  ...   

                                        movie_imdb_link num_user_for_reviews  \
0     http://www.imdb.com/title/tt0499549/?ref_=fn_t...             3,054.00   
1     http://www.imdb.com/title/tt0449088/?ref_=fn_t...             1,238.00   
2     http://www.imdb.com/title/tt2379713/?ref_=fn_t...               994.00   
3     http://www.imdb.com/title/tt1345836/?ref_=fn_t...             2,701.00   
5     http://www.imdb.com/title/tt0401729/?ref_=fn_t...               738.00   
...                                                 ...                  ...   
5033  http://www.imdb.com/title/tt0390384/?ref_=fn_t...               371.00   
5035  http://www.imdb.com/title/tt0104815/?ref_=fn_t...               130.00   
5037  http://www.imdb.com/title/tt1880418/?ref_=fn_t...                14.00   
5041  http://www.imdb.com/title/tt2070597/?ref_=fn_t...                 9.00   
5042  http://www.imdb.com/title/tt0378407/?ref_=fn_t...                84.00   

      language  country content_rating     budget_usd title_year  \
0      En

Now, let's take a look at how our cleaning is going.

In [38]:
# Check again to see how many missing values we have
data.isna().sum()

color                         1
director_name                 0
num_critic_for_reviews        3
duration_mins                 1
director_facebook_likes       0
actor_3_facebook_likes        9
actor_2_name                  3
actor_1_facebook_likes        2
gross_usd                     0
genres                        0
actor_1_name                  2
movie_title                   0
num_voted_users               0
cast_total_facebook_likes     0
actor_3_name                  9
facenumber_in_poster          6
plot_keywords                25
movie_imdb_link               0
num_user_for_reviews          1
language                      3
country                       0
content_rating               36
budget_usd                    0
title_year                    0
actor_2_facebook_likes        3
imdb_score                    0
movie_facebook_likes          0
dtype: int64

We have very few missing values remaining -- if we had more time, we could address each individually, but we don't.  Let's drop all remaining rows that have missing values.

In [39]:
print(data.shape)
# Dropping all remaining rows that have null values
data.dropna(axis=0, inplace=True)
print(data.shape)

(3625, 27)
(3555, 27)


## Feature Engineering
Sometimes you might want to use your knowledge of the subject to create new features to help you make predictions

Here, I think that the lead actor might have an influence on the gross revenue of a film, so let's check those out to see if we can turn these into a numerical feature for modeling.

In [41]:
# Check out the actors with the most movies in the set
data['actor_1_name'].value_counts()

Robert De Niro        41
Johnny Depp           34
Denzel Washington     29
J.K. Simmons          29
Matt Damon            28
                      ..
Olivia Newton-John     1
Matt Shively           1
Robert John Burke      1
James Martin Kelly     1
Emmanuelle Seigner     1
Name: actor_1_name, Length: 1371, dtype: int64

The actors with the most lead roles have 20+ lead roles in this data set.  Let's make a feature that includes whether or not the lead actor has starred in a lot of movies (20+).

For this, we will use a list comprehension, which is a concise way to create a list.  The basic syntax for this is:
- `[expression if conditional for item in list]`
  - where the `expression` is based on an item in the list, 
  - `if conditional` filters down the list
  - `for item in list` breaks down a list into individual items. 
 

In [42]:
# List comprehension example
example_list = [1, 2, 3, 4]
[number for number in example_list if number % 2 == 0]

[2, 4]

In [43]:
# Identify all movie counts, select all star actors
lead_movie_counts = data['actor_1_name'].value_counts()

In [44]:
lead_movie_counts

Robert De Niro        41
Johnny Depp           34
Denzel Washington     29
J.K. Simmons          29
Matt Damon            28
                      ..
Olivia Newton-John     1
Matt Shively           1
Robert John Burke      1
James Martin Kelly     1
Emmanuelle Seigner     1
Name: actor_1_name, Length: 1371, dtype: int64

In [45]:

star_actors = lead_movie_counts[lead_movie_counts>=20].index


In [46]:
star_actors

Index(['Robert De Niro', 'Johnny Depp', 'Denzel Washington', 'J.K. Simmons',
       'Matt Damon', 'Nicolas Cage', 'Robert Downey Jr.', 'Steve Buscemi',
       'Robin Williams', 'Bruce Willis', 'Bill Murray', 'Harrison Ford',
       'Jason Statham', 'Liam Neeson', 'Tom Hanks', 'Keanu Reeves',
       'Tom Cruise', 'Gerard Butler', 'Kevin Spacey', 'Will Ferrell',
       'Morgan Freeman'],
      dtype='object')

In [47]:
# Set `lead_star` = 1 if actor is in star_actors, otherwise 0
data['lead_star'] = [1 if x in star_actors else 0 for x in data['actor_1_name']]
data.head()


color      director_name  num_critic_for_reviews  duration_mins  \
0  Color      James Cameron                  723.00         178.00   
1  Color     Gore Verbinski                  302.00         169.00   
2  Color         Sam Mendes                  602.00         148.00   
3  Color  Christopher Nolan                  813.00         164.00   
5  Color     Andrew Stanton                  462.00         132.00   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                     0.00                  855.00  Joel David Moore   
1                   563.00                1,000.00     Orlando Bloom   
2                     0.00                  161.00      Rory Kinnear   
3                22,000.00               23,000.00    Christian Bale   
5                   475.00                  530.00   Samantha Morton   

   actor_1_facebook_likes      gross_usd                           genres  \
0                1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1               40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2               11,000.00 200,074,175.00        Action|Adventure|Thriller   
3               27,000.00 448,130,642.00                  Action|Thriller   
5                  640.00  73,058,679.00          Action|Adventure|Sci-Fi   

   ... num_user_for_reviews language  country  content_rating     budget_usd  \
0  ...             3,054.00  English      USA           PG-13 237,000,000.00   
1  ...             1,238.00  English      USA           PG-13 300,000,000.00   
2  ...               994.00  English       UK           PG-13 245,000,000.00   
3  ...             2,701.00  English      USA           PG-13 250,000,000.00   
5  ...               738.00  English      USA           PG-13 263,700,000.00   

   title_year actor_2_facebook_likes imdb_score  movie_facebook_likes  \
0        2009                 936.00       7.90                 33000   
1        2007               5,000.00       7.10                     0   
2        2015                 393.00       6.80                 85000   
3        2012              23,000.00       8.50                164000   
5        2012                 632.00       6.60                 24000   

  lead_star  
0         0  
1         1  
2         0  
3         0  
5         0  

[5 rows x 28 columns]

Perhaps the `content_rating` will have an impact on the gross revenue of the movie.  We already have this data, but our models will want the information in numerical form, not in words.  

To do this, we need to encode the ratings as numbers.  We can use the pandas [`get_dummies`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html) function.

In [49]:
# Encoding ratings as dummy variables
content_ratings = pd.get_dummies(data['content_rating'])
content_ratings

Approved  G  GP  M  NC-17  Not Rated  PG  PG-13  Passed  R  Unrated  X
0            0  0   0  0      0          0   0      1       0  0        0  0
1            0  0   0  0      0          0   0      1       0  0        0  0
2            0  0   0  0      0          0   0      1       0  0        0  0
3            0  0   0  0      0          0   0      1       0  0        0  0
5            0  0   0  0      0          0   0      1       0  0        0  0
...        ... ..  .. ..    ...        ...  ..    ...     ... ..      ... ..
5026         0  0   0  0      0          0   0      0       0  1        0  0
5033         0  0   0  0      0          0   0      1       0  0        0  0
5035         0  0   0  0      0          0   0      0       0  1        0  0
5037         0  0   0  0      0          1   0      0       0  0        0  0
5042         0  0   0  0      0          0   1      0       0  0        0  0

[3555 rows x 12 columns]

In [50]:
# Merge the encoded data back on to the original data
data = data.join(content_ratings)
data.head(3)

color   director_name  num_critic_for_reviews  duration_mins  \
0  Color   James Cameron                  723.00         178.00   
1  Color  Gore Verbinski                  302.00         169.00   
2  Color      Sam Mendes                  602.00         148.00   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                     0.00                  855.00  Joel David Moore   
1                   563.00                1,000.00     Orlando Bloom   
2                     0.00                  161.00      Rory Kinnear   

   actor_1_facebook_likes      gross_usd                           genres  \
0                1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1               40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2               11,000.00 200,074,175.00        Action|Adventure|Thriller   

   ... GP  M  NC-17  Not Rated PG  PG-13 Passed  R  Unrated  X  
0  ...  0  0      0          0  0      1      0  0        0  0  
1  ...  0  0      0          0  0      1      0  0        0  0  
2  ...  0  0      0          0  0      1      0  0        0  0  

[3 rows x 40 columns]

Now, we could do a similar type of encoding for any of the other categorical variables, but encoding all of the actor names would create too many columns.  We'll stick with the numerical data for now and select them using the Pandas [`select_dtypes`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html) function.

In [51]:
# Select columns by data type - number
numerical_data = data.select_dtypes(include='number')
numerical_data.head()

num_critic_for_reviews  duration_mins  director_facebook_likes  \
0                  723.00         178.00                     0.00   
1                  302.00         169.00                   563.00   
2                  602.00         148.00                     0.00   
3                  813.00         164.00                22,000.00   
5                  462.00         132.00                   475.00   

   actor_3_facebook_likes  actor_1_facebook_likes      gross_usd  \
0                  855.00                1,000.00 760,505,847.00   
1                1,000.00               40,000.00 309,404,152.00   
2                  161.00               11,000.00 200,074,175.00   
3               23,000.00               27,000.00 448,130,642.00   
5                  530.00                  640.00  73,058,679.00   

   num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
0           886204                       4834                  0.00   
1           471220                      48350                  0.00   
2           275868                      11700                  1.00   
3          1144337                     106759                  0.00   
5           212204                       1873                  1.00   

   num_user_for_reviews  ...  GP  M  NC-17  Not Rated  PG  PG-13  Passed  R  \
0              3,054.00  ...   0  0      0          0   0      1       0  0   
1              1,238.00  ...   0  0      0          0   0      1       0  0   
2                994.00  ...   0  0      0          0   0      1       0  0   
3              2,701.00  ...   0  0      0          0   0      1       0  0   
5                738.00  ...   0  0      0          0   0      1       0  0   

   Unrated  X  
0        0  0  
1        0  0  
2        0  0  
3        0  0  
5        0  0  

[5 rows x 28 columns]

In [52]:
numerical_data = data.select_dtypes(include=['number','object'])
numerical_data.head()

color      director_name  num_critic_for_reviews  duration_mins  \
0  Color      James Cameron                  723.00         178.00   
1  Color     Gore Verbinski                  302.00         169.00   
2  Color         Sam Mendes                  602.00         148.00   
3  Color  Christopher Nolan                  813.00         164.00   
5  Color     Andrew Stanton                  462.00         132.00   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                     0.00                  855.00  Joel David Moore   
1                   563.00                1,000.00     Orlando Bloom   
2                     0.00                  161.00      Rory Kinnear   
3                22,000.00               23,000.00    Christian Bale   
5                   475.00                  530.00   Samantha Morton   

   actor_1_facebook_likes      gross_usd                           genres  \
0                1,000.00 760,505,847.00  Action|Adventure|Fantasy|Sci-Fi   
1               40,000.00 309,404,152.00         Action|Adventure|Fantasy   
2               11,000.00 200,074,175.00        Action|Adventure|Thriller   
3               27,000.00 448,130,642.00                  Action|Thriller   
5                  640.00  73,058,679.00          Action|Adventure|Sci-Fi   

   ... GP  M  NC-17  Not Rated PG  PG-13 Passed  R  Unrated  X  
0  ...  0  0      0          0  0      1      0  0        0  0  
1  ...  0  0      0          0  0      1      0  0        0  0  
2  ...  0  0      0          0  0      1      0  0        0  0  
3  ...  0  0      0          0  0      1      0  0        0  0  
5  ...  0  0      0          0  0      1      0  0        0  0  

[5 rows x 40 columns]